### PR TITLE
Add classes Status and StatusOr, derived from protobuf code

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: base/status.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+class Status {
+ public:
+  // Creates a "successful" status.
+  Status();
+
+  Status(Error const error, std::string const& message);
+  Status(Status const&);
+  Status& operator=(Status const& x);
```
Comment (eggrobin) [inline base/status.hpp]: Can we `default` the copy constructor and assignment? Also add `default` move constructor and assignment.
In fact, why aren't they implicitly declared? `Status` seems to contain only copyable and movable things.

File: base/status.hpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+  // Some pre-defined Status objects
```
Comment (eggrobin) [inline base/status.hpp]: Full stop at the end of a comment.

File: base/status.hpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  static const Status OK;
+  static const Status CANCELLED;
+  static const Status UNKNOWN;
+
+  // Accessor
```
Comment (eggrobin) [inline base/status.hpp]: Accessors.

File: base/statusor.hpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  //
+  // REQUIRES: if T is a plain pointer, value != NULL. This requirement is
+  // DCHECKed. In optimized builds, passing a NULL pointer here will have
+  // the effect of passing PosixErrorSpace::EINVAL as a fallback.
+  StatusOr(T const& value);
+
+  // Copy constructor.
+  StatusOr(StatusOr const& other);
+
+  // Conversion copy constructor, T must be copy constructible from U
```
Comment (eggrobin) [inline base/statusor.hpp]: Full stop

File: base/statusor.hpp [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  template<typename U>
+  StatusOr(StatusOr<U> const& other);
+
+  // Assignment operator.
+  StatusOr& operator=(StatusOr const& other);
+
+  // Conversion assignment operator, T must be assignable from U
```
Comment (eggrobin) [inline base/statusor.hpp]: Full stop

File: base/statusor.hpp [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+  template<typename U>
+  StatusOr& operator=(StatusOr<U> const& other);
+
+  // Returns a reference to our status. If this contains a T, then
+  // returns Status::OK.
+  Status const& status() const;
+
+  // Returns this->status().ok()
```
Comment (eggrobin) [inline base/statusor.hpp]: Full stop

File: base/statusor.hpp [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+  // so it is convenient and sensible to be able to do 'return T()'
+  // when when the return type is StatusOr<T>.
+  //
+  // REQUIRES: if T is a plain pointer, value != NULL. This requirement is
+  // DCHECKed. In optimized builds, passing a NULL pointer here will have
+  // the effect of passing PosixErrorSpace::EINVAL as a fallback.
+  StatusOr(T const& value);
+
+  // Copy constructor.
+  StatusOr(StatusOr const& other);
```
Comment (eggrobin) [inline base/statusor.hpp]: This should probably be `default`ed, and if we want to delete the assignment, I'd rather do that explicitly too.

File: base/statusor.hpp [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+#include <string>
+#include <utility>
+
+#include "base/status.hpp"
+
+namespace principia {
+namespace base {
+
+template<typename T>
+class StatusOr {
```
Comment (eggrobin) [inline base/statusor.hpp]: In its current state, even disregarding move semantics, this class does not seem to be adapted to our needs:
- It requires that `T` be default-constructible, this is rarely the case (an obvious example is returning `not_null`, but even `KeplerOrbit`, the linear maps, or `RigidMotion` are devoid of default constructors).
- It tries to replicate some of the functionality of `not_null`.

File: base/status.hpp (hunk 1 above)
Comment (pleroy) [inline base/status.hpp]: Done.

File: base/status.hpp (hunk 2 above)
Comment (pleroy) [inline base/status.hpp]: Done.

File: base/status.hpp (hunk 3 above)
Comment (pleroy) [inline base/status.hpp]: Done.

File: base/statusor.hpp (hunk 4 above)
Comment (pleroy) [inline base/statusor.hpp]: Done.

File: base/statusor.hpp (hunk 5 above)
Comment (pleroy) [inline base/statusor.hpp]: Done.

File: base/statusor.hpp (hunk 6 above)
Comment (pleroy) [inline base/statusor.hpp]: Done.

File: base/statusor.hpp (hunk 7 above)
Comment (pleroy) [inline base/statusor.hpp]: Done.

File: base/statusor.hpp (hunk 8 above)
Comment (pleroy) [inline base/statusor.hpp]: Done.

File: base/statusor.hpp [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// StatusOr<T> is the union of a Status object and a T
+// object. StatusOr models the concept of an object that is either a
+// usable value, or an error Status explaining why such a value is
+// not present. To this end, StatusOr<T> does not allow its Status
+// value to be Status::OK. Further, StatusOr<T*> does not allow the
+// contained pointer to be NULL.
```
Comment (eggrobin) [inline base/statusor.hpp]: These comments need to be updated.

File: base/statusor.hpp [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  // value, so it is convenient and sensible to be able to do 'return
+  // Status()' when the return type is StatusOr<T>.
+  //
+  // REQUIRES: status != Status::OK. This requirement is DCHECKed.
+  // In optimized builds, passing Status::OK here will have the effect
+  // of passing PosixErrorSpace::EINVAL as a fallback.
+  StatusOr(Status const& status);  // NOLINT
+
+  // Construct a new StatusOr with the given value. If T is a plain pointer,
+  // value must not be NULL. After calling this constructor, calls to
```
Comment (eggrobin) [inline base/statusor.hpp]: No special treatment for pointers now.

File: base/statusor.hpp [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  // Returns a reference to our status. If this contains a T, then
+  // returns Status::OK.
+  Status const& status() const;
+
+  // Returns this->status().ok().
+  bool ok() const;
+
+  // Returns a reference to our current value, or CHECK-fails if !this->ok().
+  // If you need to initialize a T object from the stored value,
+  // ConsumeValueOrDie() may be more efficient.
```
Comment (eggrobin) [inline base/statusor.hpp]: `ConsumeValueOrDie` may be more efficient, but it's not declared.

**Comment by eggrobin** [comment]
Could we use `|vertical_lines|` for code in the comments, as we do in the rest of the codebase?
Things like
```cpp
 // Returns this->status().ok().
```
look a little weird.

File: base/statusor.hpp (hunk 9 above)
Comment (pleroy) [inline base/statusor.hpp]: Done.

File: base/statusor.hpp (hunk 11 above)
Comment (pleroy) [inline base/statusor.hpp]: Done.

File: base/statusor.hpp (hunk 10 above)
Comment (pleroy) [inline base/statusor.hpp]: Done.

File: base/statusor.hpp [hunk 12]
```diff
@@ -102,51 +101,45 @@ class StatusOr {
   // Construct a new StatusOr with Status::UNKNOWN status
   StatusOr();
 
-  // Construct a new StatusOr with the given non-ok status. After calling
-  // this constructor, calls to ValueOrDie() will CHECK-fail.
+  // Construct a new object with the given non-ok status. After calling
+  // this constructor, calls to |ValueOrDie()| will-fail.
```
Comment (eggrobin) [inline base/statusor.hpp]: will-fail

File: base/statusor.hpp [hunk 13]
```diff
@@ -102,51 +101,45 @@ class StatusOr {
...
+  // NOTE: Not explicit - we want to use |StatusOr<T>| as a return
+  // value, so it is convenient and sensible to be able to do |return Status();|
+  // when the return type is |StatusOr<T>|.
   StatusOr(Status const& status);  // NOLINT
 
-  // Construct a new StatusOr with the given value. If T is a plain pointer,
-  // value must not be NULL. After calling this constructor, calls to
-  // ValueOrDie() will succeed, and calls to status() will return OK.
+  // Construct a new StatusOr with the given value. After calling this
+  // constructor, calls to |ValueOrDie()| will succeed, and calls to status()
```
Comment (eggrobin) [inline base/statusor.hpp]: `status()`

File: base/statusor.hpp [hunk 14]
```diff
@@ -102,51 +101,45 @@ class StatusOr {
...
-  // of passing PosixErrorSpace::EINVAL as a fallback.
+  // NOTE: Not explicit - we want to use |StatusOr<T>| as a return
+  // value, so it is convenient and sensible to be able to do |return Status();|
+  // when the return type is |StatusOr<T>|.
   StatusOr(Status const& status);  // NOLINT
 
-  // Construct a new StatusOr with the given value. If T is a plain pointer,
-  // value must not be NULL. After calling this constructor, calls to
-  // ValueOrDie() will succeed, and calls to status() will return OK.
+  // Construct a new StatusOr with the given value. After calling this
```
Comment (eggrobin) [inline base/statusor.hpp]: `StatusOr`

File: base/statusor.hpp [hunk 15]
```diff
@@ -102,51 +101,45 @@ class StatusOr {
...
-  // Conversion copy constructor, T must be copy constructible from U.
+  // Conversion copy constructor, |T| must be copy constructible from |U|.
   template<typename U>
   StatusOr(StatusOr<U> const& other);
 
   // Assignment operator.
   StatusOr& operator=(StatusOr const& other) = default;
 
-  // Conversion assignment operator, T must be assignable from U.
+  // Conversion assignment operator, |T| must be assignable from |U|.
```
Comment (eggrobin) [inline base/statusor.hpp]: copy-assignable

File: base/statusor.hpp (hunk 12 above)
Comment (pleroy) [inline base/statusor.hpp]: will.i.am

**Pull request merged**